src/visual.ts

```diff
@@ -983,43 +984,60 @@ module powerbi.extensibility.visual {
             let yPos: number = 0,
                 yDiff: number = Timeline.DefaultYDiff;
 
-            this.renderLabels(
-                extendedLabels.yearLabels,
-                this.yearLabelsSelection,
-                yPos,
-                granularityType === 0);
-
-            yPos += yDiff;
-
-            this.renderLabels(
-                extendedLabels.quarterLabels,
-                this.quarterLabelsSelection,
-                yPos,
-                granularityType === 1);
-
-            yPos += yDiff;
-
-            this.renderLabels(
-                extendedLabels.monthLabels,
-                this.monthLabelsSelection,
-                yPos,
-                granularityType === 2);
+            // Removing currently displayed labels
+            this.mainGroupSelection
+            .selectAll(Timeline.TimelineSelectors.textLabel.selectorName)
+            .remove();
+
+            if (timelineSettings.labels.displayAll || granularityType == GranularityType.year) {
+                this.renderLabels(
+                    extendedLabels.yearLabels,
+                    this.yearLabelsSelection,
+                    yPos,
+                    granularityType === 0);
+                if (timelineSettings.labels.displayAll)
```

Comment:
Please use - If() { } construction. And I guess you can move this line to before call of rendecCell  function(line 1042). This code is repeated